File: app/src/main/java/com/demo/railbridge/sdk/ScenarioController.java
package com.demo.railbridge.sdk;

import java.util.concurrent.atomic.AtomicReference;

public class ScenarioController {

    private final AtomicReference<ScenarioPreset> activePreset;

    public ScenarioController() {
        this(ScenarioPreset.NORMAL);
    }

    public ScenarioController(ScenarioPreset initialPreset) {
        activePreset = new AtomicReference<>(sanitize(initialPreset));
    }

    public ScenarioPreset getActivePreset() {
        return activePreset.get();
    }

    public void setActivePreset(ScenarioPreset preset) {
        activePreset.set(sanitize(preset));
    }

    public void setActivePreset(String presetValue) {
        setActivePreset(ScenarioPreset.fromValue(presetValue));
    }

    private ScenarioPreset sanitize(ScenarioPreset preset) {
        return preset == null ? ScenarioPreset.NORMAL : preset;
    }
}

File: ios/RailBridgeIOS/RailBridgeIOS/SDK/ScenarioController.swift
import Foundation

final class ScenarioController {
    private let queue = DispatchQueue(label: "com.demo.railbridge.ios.scenario")
    private var activePresetStorage: ScenarioPreset = .normal

    var activePreset: ScenarioPreset {
        queue.sync { activePresetStorage }
    }

    func setActivePreset(_ rawValue: String) {
        let preset = ScenarioPreset.from(rawValue)
        queue.sync {
            activePresetStorage = preset
        }
    }

    func reset() {
        setActivePreset(ScenarioPreset.normal.rawValue)
    }
}
